src/v/kafka/requests/fetch_request.cc

```diff
@@ -449,41 +476,16 @@ fetch_api::process(request_context&& rctx, ss::smp_service_group ssg) {
         // top-level error is used for session-level errors, but we do not yet
         // implement session management.
         octx.response.error = error_code::none;
-        return fetch_topic_partitions(octx).then([&octx] {
-            /*
-             * fast out
-             *
-             * response_size is a size_t accumulated value that we track as we
-             * build the response. all of the kafka types (e.g. min_bytes) are
-             * annoyingly signed.
-             */
-            if (
-              !octx.request.debounce_delay()
-              || (int32_t)octx.response_size >= octx.request.min_bytes
-              || octx.request.topics.empty() || octx.response_error) {
-                return octx.rctx.respond(std::move(octx.response));
-            }
-
-            /*
-             * debounce since not enough bytes were collected.
-             *
-             * TODO:
-             *   - actual debouncing would collect additional data if
-             * available, but this only introduces the delay. the delay is
-             * still important for now so that the client and server do
-             * not sit in tight req/rep loop, and once the client gets an
-             * ack it can retry. the implementation of debouncing should
-             * be done or at least coordinated with the storage layer.
-             * Otherwise we'd have to developer heuristics on top of
-             * storage for polling.
-             *
-             *   - this needs to be abortable sleep coordinated with
-             * server shutdown.
-             */
-            auto delay = octx.deadline - model::timeout_clock::now();
-            return ss::sleep<model::timeout_clock>(delay).then(
-              [&octx] { return octx.rctx.respond(std::move(octx.response)); });
-        });
+        // first fetch, do not wait
+        return fetch_topic_partitions(octx)
+          .then([&octx] {
+              octx.initial_fetch = false;
+              return ss::do_until(
+                [&octx] { return octx.should_stop_fetch(); },
+                [&octx] { return fetch_topic_partitions(octx); });
```

Comment:
@mmaslankaprv it looks like this may result in a `fetch_response` containing more partition entries for a topic than had been requested if a retry occurs. (a) is that true (b) is that compliant with the kafka protocol?

Reply:
If I understand correctly partition list is contained in the request and that is constant, we do not query for more partitions

Reply:
if `fetch_topic_partitions` is repeated in this loop, it looks like `octx.start_response_topic` will add a duplicate entry in `fetch_response::partitions`.